### PR TITLE
On success buttons

### PR DESCRIPTION
**Pull request by slavik-lvovsky**

### CODE REVIEW

File: backend/src/yeomanui.ts [hunk 1]
```diff
@@ -102,7 +103,7 @@ export class YeomanUI {
     entry.set(methodName, handler);
   }
 
-  private getCustomQuestionEventHandler(questionType: string, methodName: string): Function {
+  public getCustomQuestionEventHandler(questionType: string, methodName: string): Function {
```
Comment (perspectivus1) [inline backend/src/yeomanui.ts:106]: Why change to public?

Comment (slavik-lvovsky) [reply backend/src/yeomanui.ts:106]: Mistake.

File: frontend/src/App.vue [hunk 2]
```diff
@@ -381,6 +385,7 @@ export default {
         "setPromptList",
         "generatorInstall",
         "generatorDone",
+        // "selectTargetFolder",
```
[the remark this answers is not in the document]
Comment (slavik-lvovsky) [reply frontend/src/App.vue:388]: It is preparation for my BLI.

File: frontend/src/App.vue [hunk 3]
```diff
@@ -170,11 +171,14 @@ export default {
     }
   },
   methods: {
+    // selectTargetFolder() {
+    //   this.rpc.invoke("selectTargetFolder", []).then(result => {
+    //     this.rpc.invoke("logError", [result]);
+    //   });
+    // },
```
[the remark this answers is not in the document]
Comment (slavik-lvovsky) [reply frontend/src/App.vue:178]: It is preparation for my BLI.

File: backend/src/yeomanui.ts [hunk 4]
```diff
@@ -228,6 +227,14 @@ export class YeomanUI {
...
   }
 
+  // public async selectTargetFolder() {
+  //   try {
+  //     return await this.getCustomQuestionEventHandler("folder-browser", "getPath")();
+  //   } catch (error) {
+  //     return YeomanUI.PROJECTS;
+  //   }
+  // }
+
```
[the remark this answers is not in the document]
Comment (slavik-lvovsky) [reply backend/src/yeomanui.ts:237]: It is preparation for my BLI.

File: backend/src/vscode-youi-events.ts [hunk 5]
```diff
@@ -48,20 +50,36 @@ export class VSCodeYouiEvents implements YouiEvents {
...
-            const OpenWorkspace = 'Open in New Workspace';
-            const AddToWorkspace = (vscode.workspace.workspaceFolders && vscode.workspace.workspaceFolders.length) ? 'Add to Workspace' : undefined;
-            vscode.window.showInformationMessage('The project has been successfully generated.\nWhat would you like to do with it?', AddToWorkspace, OpenWorkspace).then(selection => {
-                if (selection === OpenWorkspace) {
+            let addToWorkspaceButton: any;
+            let openInNewWorkspaceButton: any;
+            let infoMessage = "The project has been successfully generated.";
+            const uriTargetFolder = vscode.Uri.file(targetPath);
+
+            if (this.genFilter.type !== GeneratorType.module) {
```
[the remark this answers is not in the document]
Comment (slavik-lvovsky) [reply backend/src/vscode-youi-events.ts:62]: No, We decided with ido and Avital that Projecta and All are relevant

File: backend/src/vscode-youi-events.ts [hunk 6]
```diff
@@ -1,20 +1,22 @@
...
     public static installing: boolean;
 
-    constructor(rpc : RpcCommon, webviewPanel: vscode.WebviewPanel) {
+    constructor(rpc : RpcCommon, webviewPanel: vscode.WebviewPanel, private genFilter: GeneratorFilter) {
         this.rpc = rpc; 
         this.webviewPanel = webviewPanel;       
     }
 
-    public doGeneratorDone(success: boolean, message: string, targetPath = ""): void {
+    public doGeneratorDone(generationSucceeded: boolean, message: string, targetPath: string = ""): void {
```
[the remark this answers is not in the document]
Comment (slavik-lvovsky) [reply backend/src/vscode-youi-events.ts:17]: I took me time to understand what is the meaning of success that is why i renamed it.

File: backend/src/vscode-youi-events.ts [hunk 7]
```diff
@@ -1,20 +1,22 @@
...
 import { RpcCommon } from "@sap-devx/webview-rpc/out.ext/rpc-common";
+import { GeneratorFilter, GeneratorType } from './filter';
 
 export class VSCodeYouiEvents implements YouiEvents {
     private rpc: RpcCommon;
     private webviewPanel: vscode.WebviewPanel;
     public static installing: boolean;
 
-    constructor(rpc : RpcCommon, webviewPanel: vscode.WebviewPanel) {
+    constructor(rpc : RpcCommon, webviewPanel: vscode.WebviewPanel, private genFilter: GeneratorFilter) {
```
[the remark this answers is not in the document]
Comment (slavik-lvovsky) [reply backend/src/vscode-youi-events.ts:12]: Done.

File: backend/src/vscode-youi-events.ts [hunk 8]
```diff
@@ -48,20 +50,36 @@ export class VSCodeYouiEvents implements YouiEvents {
...
-            const AddToWorkspace = (vscode.workspace.workspaceFolders && vscode.workspace.workspaceFolders.length) ? 'Add to Workspace' : undefined;
-            vscode.window.showInformationMessage('The project has been successfully generated.\nWhat would you like to do with it?', AddToWorkspace, OpenWorkspace).then(selection => {
-                if (selection === OpenWorkspace) {
+            let addToWorkspaceButton: any;
+            let openInNewWorkspaceButton: any;
+            let infoMessage = "The project has been successfully generated.";
+            const uriTargetFolder = vscode.Uri.file(targetPath);
+
+            if (this.genFilter.type !== GeneratorType.module) {
+                const targetFolderInWorkspace: vscode.WorkspaceFolder = vscode.workspace.getWorkspaceFolder(uriTargetFolder);
```
[the remark this answers is not in the document]
Comment (slavik-lvovsky) [reply backend/src/vscode-youi-events.ts:63]: Done

File: backend/src/yeomanui.ts [hunk 9]
```diff
@@ -82,6 +82,7 @@ export class YeomanUI {
     this.rpc.registerMethod({ func: this.toggleOutput, thisArg: this });
```
[the remark this answers is not in the document]
Comment (slavik-lvovsky) [reply backend/src/yeomanui.ts:82]: No, we need this param here, We will get it from extension once i finish my BLI

File: backend/src/yeomanui.ts (hunk 4 above)
[the remark this answers is not in the document]
Comment (tomer-epstein) [reply backend/src/yeomanui.ts:237]: save it in notepad++ ;-)

File: frontend/src/App.vue (hunk 3 above)
[the remark this answers is not in the document]
Comment (tomer-epstein) [reply frontend/src/App.vue:178]: save it in notepad++ ;-) or in a different branch.

File: backend/src/vscode-youi-events.ts [hunk 10]
```diff
@@ -56,26 +56,31 @@ export class VSCodeYouiEvents implements YouiEvents {
         VSCodeYouiEvents.installing = false;
         
         if (success) {
-            let addToWorkspaceButton: any;
-            let openInNewWorkspaceButton: any;
+            const addToWorkspace: string = "Add to Workspace";
+            const openInNewWorkspace: any = "Open in New Workspace";
+            const items: string[] = [];
```
Comment (tomer-epstein) [inline backend/src/vscode-youi-events.ts:61]: I think thi doesn't work in theia. keep passing undefined,undefined.

Comment (slavik-lvovsky) [reply backend/src/vscode-youi-events.ts:61]: It does not work with single or double undefined too

Comment (slavik-lvovsky) [reply backend/src/vscode-youi-events.ts:61]: I check in theia and both or single buttons are displayed
The only problem is messages without buttons

File: backend/src/vscode-youi-events.ts [hunk 11]
```diff
@@ -56,26 +56,31 @@ export class VSCodeYouiEvents implements YouiEvents {
...
+                if (!_.isEmpty(items)) {
                     infoMessage = infoMessage.concat("\nWhat would you like to do with it?");
                 } 
             }
                 
-            vscode.window.showInformationMessage(infoMessage, addToWorkspaceButton, openInNewWorkspaceButton).then(selection => {
-                if (selection === openInNewWorkspaceButton) {
-                    this.executeCommand("vscode.openFolder", targetPath);
-                } else if (selection === addToWorkspaceButton) {
+            vscode.window.showInformationMessage(infoMessage, ...items).then(selection => {
```
Comment (tomer-epstein) [inline backend/src/vscode-youi-events.ts]: vscode.window.showInformationMessage(infoMessage, addToWorkspace, openInNewWorkspace).then(selection => {

File: backend/src/vscode-youi-events.ts (hunk 10 above)
Comment (tomer-epstein) [reply backend/src/vscode-youi-events.ts:61]: heelo world message must work. in Theia. We can check it with simple vscode ext only showing message. if not working then we can open issue to theia.

Comment (slavik-lvovsky) [reply backend/src/vscode-youi-events.ts:61]: it does not work, i have already checked

Comment (slavik-lvovsky) [reply backend/src/vscode-youi-events.ts:61]: or might be it does not work only in this environment

Comment (tomer-epstein) [reply backend/src/vscode-youi-events.ts:61]: clone https://github.com/tomer-epstein/vscode-add-Root-Folder.git

change: vscode.commands.executeCommand("workbench.action.addRootFolder");
to: vscode.window.showInformationMessage('Hello World!'); 

it works.

image

**Comment by tomer-epstein** [comment]
now it doesn't create the new project under the current open WS, is this as designed?

File: backend/src/vscode-youi-events.ts (hunk 10 above)
Comment (tomer-epstein) [reply backend/src/vscode-youi-events.ts:61]: works good - 
both
const items: string[] = [];
and
vscode.window.showInformationMessage(infoMessage, ...items)

File: backend/src/vscode-youi-events.ts [hunk 12]
```diff
@@ -48,31 +52,45 @@ export class VSCodeYouiEvents implements YouiEvents {
         });
     }
 
-    private showDoneMessage(success: boolean, message: string, targetPath: string): void {
+    private showDoneMessage(success: boolean, errorMmessage: string, targetPath: string): Thenable<any> {
```
Comment (tomer-epstein) [inline backend/src/vscode-youi-events.ts:55]: here it's - success: boolean
at doGeneratorDone it's generationSucceeded: boolean
I prefer success in both

File: backend/src/yeomanui.ts [hunk 13]
```diff
@@ -305,7 +303,8 @@ export class YeomanUI {
   private setGenInstall(gen: any) {
     const originalPrototype = Object.getPrototypeOf(gen);
     const originalGenInstall = _.get(originalPrototype, "install");
-    if (originalGenInstall) {
+    if (originalGenInstall && !originalPrototype._uiInstall) {
+      originalPrototype._uiInstall = true;
```
Comment (tomer-epstein) [inline backend/src/yeomanui.ts:307]: Nice

File: backend/src/vscode-youi-events.ts (hunk 12 above)
Comment (slavik-lvovsky) [reply backend/src/vscode-youi-events.ts:55]: Done

**Comment by slavik-lvovsky** [comment]
> now it doesn't create the new project under the current open WS, is this as designed?

What do you mean ?

**Comment by tomer-epstein** [comment]
> > now it doesn't create the new project under the current open WS, is this as designed?
> 
> What do you mean ?

- workspace is already opened on junkAA
- Create new foodq project junkBB

Expected - junkBB created under junkAA
Actual - junkBB created under projects folder.

** I don't remember if we decide that this is by design now? **

**Comment by slavik-lvovsky** [comment]
Now I understand ☺
I have removed the code that Sagi added.
Currently, all projects are created as before under projects.
In the BLI that I want to continue next week it will be changed.
Thank you for review.

It very hard for me to develop without stable theia environment ☹

**Comment by slavik-lvovsky** [comment]
Are there other things to fix ?

**Pull request merged**